ASoC: SOF: Intel: hda: allow operation without i915

kv2019i opened the pull request:
Add support to configure HDA controller with external codec even if iDisp codec in i915 is not available.
    
This can happen for multiple reasons:
     - internal graphics is disabled on the system
     - i915 driver is not enabled in kernel or it fails to init
     - i915 codec reports error in HDA codec probe
     - HDA codec driver probe fails
    
Address all these scenarios, but keep using the same topology configuration still. In case failures occur, HDMI PCM nodes are created, but they will report error at open. No ALSA mixer controls are created.

Fixes: #1658

Note to @plbossart : this will conflict with "ASoC: SOF: Intel: refactor i915_get/put functions" that was sent to upstream last week. In case merging get delayed, you'll see this.

kv2019i commented:
Draft only, I didn't manage to clean up fully yet, but this looks very promising. We can extend the HDA generic driver to these cases without needing a separate topology. Plus a big bonus, transient faults with i915 hw (like the recent afg problems), won't bring down the whole SOF stack, but only the HDMI part.

lyakh commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -135,11 +135,21 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
+				skl_hda_be_dai_links[i].codecs->name =
+					"snd-soc-dummy";
+				skl_hda_be_dai_links[i].codecs->dai_name =
+					"snd-soc-dummy-dai";
```
If I understand correctly, `skl_hda_be_dai_links[i].codecs` is actually an array of `skl_hda_be_dai_links[i].num_codecs` elements. So, the above actually means `skl_hda_be_dai_links[i].codecs[0]` and it assumes that `.num_codecs` is at least 1. Maybe indeed this is a valid assumption, but it would be better to use an array representation instead of just a pointer and maybe add a comment, saying that array is guaranteed to have at least one element in it?

kv2019i replied:
@lyakh Ack. I changed this for the "production version" and used the common macros instead.

kv2019i commented:
Ready for review folks! I tested all the error scenarios with specially crafted error injection code.

kv2019i commented on sound/soc/sof/intel/hda-codec.c:
```diff
@@ -113,8 +114,14 @@ static int hda_codec_probe(struct snd_sof_dev *sdev, int address,
+		if (!hda->i915_codec_enabled) {
+			dev_dbg(sdev->dev,
+				"iDisp hw present but no driver, ignoring\n");
+			return -ENOENT;
```
One thing I wondered about is should the device be unregistered here immediately in the error case. But I ended up  not doing that. snd_hdac_ext_bus_device_remove() will clean up this device as well.

kv2019i commented:
There's one checkpatch warning (shows up as a failure from Travis).
```
CHECK: Alignment should match open parenthesis
#41: FILE: sound/soc/intel/boards/skl_hda_dsp_generic.c:65:
+SND_SOC_DAILINK_DEF(dummy_codec,
+       DAILINK_COMP_ARRAY(COMP_CODEC("snd-soc-dummy", "snd-soc-dummy-dai")));
```

... but this is matching with existing usage of SND_SOC_DAILINK_DEF in ASoC, so kept it like this.

lyakh commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -135,11 +138,20 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
+			if (codec_count == 2)
+				return -EINVAL;
+			for (i = 0; i < IDISP_DAI_COUNT; i++) {
+				skl_hda_be_dai_links[i].codecs = dummy_codec;
```
I'm a bit confused now, didn't you have
```
skl_hda_be_dai_links[i].codecs->name = ...;
```
before? That would mean, that there was at least one element assigned to `.codecs`. So, now you're overwriting that? I.e. you're relying on the fact, that the original `.codecs` was allocated with `devm_kcalloc()`, so, it will be freed anyway?

kv2019i replied:
@lyakh Yeah, trying to imitatte the best modern ASoC DAI macro usages here. :) So the codecs fields are typically filled with the SND_SOC_DAILINK_() macros and that also applies to this machine driver. These macros create and fill static structs in the compilation unit. So be default the codecs field is a pointer to a static struct declated in skl_hda_dsp_common.c . What I add in this patch is another SND_SOC_DALINK_DEF that creates static struct 'dummy_codec' and in this loop I changes codecs to point to it. Based on my understanding and review of the macros, this is safe as codecs always points to one of the statically allocated structs and never to a dynamically allocated object. This code is local to the generic hda driver, so other entities will not modify skl_hda_be_dai_links directly.

But yeah, scrutiny of the logic is much welcome.

plbossart commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -135,11 +138,20 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
+			for (i = 0; i < IDISP_DAI_COUNT; i++) {
+				skl_hda_be_dai_links[i].codecs = dummy_codec;
+				skl_hda_be_dai_links[i].num_codecs =
+					ARRAY_SIZE(dummy_codec);
```
why not remove those HDMI dailinks, instead of mucking with the number of codecs?

Is this to avoid errors with the topology, where we are trying to fnd a dailink what would not exist if we removed them? And what happens if you just use dummy-codec, the user would still see an HDMI stream but the jack detection would never signal they can be used?

kv2019i replied:
@plbossart You guessed right. If we'd remove the links, then loading the topology fails. Requiring yet another topology for this case seems wasteful. 

With the dummy codec, "HDMI/DP Jack" ALSA controls are not created, so no jack detection events will be routed to user-space. The PCM devices will be created, but they will always fail at pcm open. And this kind of matches with what actually happens if ifgx is disabled... on SOF side we have the iDisp links, but the other end is just not connected.

plbossart commented on sound/soc/sof/intel/hda-codec.c:
```diff
@@ -113,8 +114,14 @@ static int hda_codec_probe(struct snd_sof_dev *sdev, int address,
+	if ((resp & 0xFFFF0000) == IDISP_VID_INTEL) {
+		if (!hda->i915_codec_enabled) {
+			dev_dbg(sdev->dev,
+				"iDisp hw present but no driver, ignoring\n");
```
it's more 'will be ignored'. There is nothing in this function that ignores the codec, you just return a negative value that will be handled in hda_codec_probe_bus

kv2019i replied:
@plbossart hmm, I'll change this. but we do skip the codec driver probe later in the function, so first part of "ignoring" is already done in this function. but yeah, the info needs to be passed to controller side as well and that is done in the parent, so point taken.

kv2019i replied:
@plbossart "will be ignored" didn't fit 80-chars, so I left ", ignoring" part out.. not adding value

plbossart commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -135,11 +138,20 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
+				skl_hda_be_dai_links[i].num_codecs =
+					ARRAY_SIZE(dummy_codec);
+			}
+		}
```
you should clarify what happens for other machine drivers? Why would we ignore HDMI only in the case of the HDaudio codec cases?

kv2019i replied:
@plbossart We probably should modify other machine drivers, but that is less critical as those are for specific devices where HDMI either is or is not part of the product configuration, so there is less expected variation. We could either go ahead and change all the other machien drivers as well, or mybe wait for @ranj063 's mfd work to reach the point where hdmi is a separate client and potentially easier to isolate.

But for the generic HDA driver, this is much more urgent. The hda generic driver is our catch-all driver that should be much more robust against variations in configuration than the other more specific machine drivers. E.g. any machine that has IGFX disabled and has DMICs, will just fail out-of-the-box with upstream kernel now without this patch.

ranj063 replied:
@kv2019i this logic seems twisted. If idisp codecs are disabled, loading a topology with idisp pipelines should fail IMO.
By loading the tplg with idsp pipelines and using dummy codecs, we're giving the wrong signal

ranj063 replied:
Alternativey, what we should do is fail gracefully ie not let the card registration fail entirely but somehow ignore the idisp pipelines in tplg with a warning

plbossart replied:
@ranj063 we don't have a good way to discard parts of the topology when the hardware described in those topology files is just not present.

I think @kv2019i choses the least invasive solution, the pipelines remain but are essentially never used since the jack detection will never fail. However there could still be a problem if someone decided to use aplay -Dhw:x,y. @kv2019i what would happen then? Underflow?

kv2019i replied:
@plbossart @ranj063 Ack, exactly. If i915 fails to initialize and you try to aplay to one of the PCMs, you will just get an error from failing hw_error and playback won't start. Due to inherent hotplug nature of HDMI/DP monitors, no application should assume that existance of PCMs indicates that a monitor is connected. The routing is not guaranteed, so this is just plain wrong logic.

If we go with the approach of using different topologies, then we end up maintaining three topologies for each platform (hdmi and some codec, only hdmi and only some codec). If we have similar availability concerns with some other codecs, do we then provide topologies for all combinations (codec1, codec2; codec1 only; codec 2 only; and all combinations of codec1 and codec2 with and without hdmi). 

I think a simpler approach is that if only difference in topology is that some of the edge leafs of the graph are not available, we are better of using a single topology and just handling enable/disable for these leaf nodes (like HDMI here). The leafs could be input or outputs. Only if the difference is such, that we would set up a different looking topology altogether (and save some DSP resource), then maintaining it separately makes sense.

kv2019i commented:
v3 uploaded: addressed Pierre's feedback to code comments, no other changes compared to v2

kv2019i commented:
@lyakh @keyonjie @ranj063 please review, this is still waiting for a second approval

lgirdwood commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -135,11 +138,20 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
 					HDAC_DAI_COUNT + i];
 		}
-	} else if (codec_count == 2 && codec_mask & IDISP_CODEC_MASK) {
+	} else if (codec_count <= 2) {
```
minor - why do we <= 2 codecs important here ?

lgirdwood commented on sound/soc/sof/intel/hda.c:
```diff
@@ -987,7 +989,7 @@ static int hda_generic_machine_select(struct snd_sof_dev *sdev)
 				 hda_mach->drv_name);
 
-			if (codec_num == 1)
+			if (codec_num == 1 && HDA_IDISP_CODEC(bus->codec_mask))
```
minor again, not following here why codec count is important. previous author should have commented this.

plbossart commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -135,11 +138,20 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
+				skl_hda_be_dai_links[i].num_codecs =
+					ARRAY_SIZE(dummy_codec);
+			}
+		}
```
agree with @lgirdwood we should probably add comments to explain the supported configurations, so that reviewers can just look at the code without having to reverse-engineer the logic.

I think we care about the three cases, yes?

a) HDMI only
b) HDAudio with no HDMI
c) HDAudio with HDMI.

kv2019i replied:
Ack @plbossart @lgirdwood . I'll make it clearer we handle these three specific cases.

kv2019i commented:
I will let https://github.com/thesofproject/linux/pull/1764 go in first and I will then rebase this one.

plbossart commented:
> I will let #1764 go in first and I will then rebase this one.

#1764 merged

kv2019i commented:
I've addressed the comments in my local branch, but will wait and not upload until https://github.com/thesofproject/linux/pull/1767 is merged via upstream (there's a conflict).

kv2019i commented:
One remaining issues with SOF module load/unload with this patchset and latest sof-dev, will debug this today.

kv2019i commented:
@lgirdwood @plbossart Now updated to address your comments.

plbossart commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -114,13 +117,18 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
+		return -EINVAL;
 
-	if (codec_count == 1 && codec_mask & IDISP_CODEC_MASK) {
+	if (codec_count == idisp_count) {
```
Humm, confused...

idisp_count is either 0 or 0x4.

did you mean idisp_count = hweight_long(codec_mask & IDISP_CODEC_MASK) ?

kv2019i replied:
@plbossart Gah :(, thanks Pierre for catching this. I also messed up in testing as I thought I had covered this case, but I didn't as this is obviously wrong. Going back, the checks still looked too complicated as well. I now updated the 3rd try with s/disp_count/idisp_mask/ and added more comments to further clarify that we are handling two separate cases of topologies here. Hopefully now clear. Tested scenarios on CFL as well.

ranj063 commented on sound/soc/intel/boards/skl_hda_dsp_generic.c:
```diff
@@ -135,13 +144,19 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
 		card->num_dapm_widgets = ARRAY_SIZE(skl_hda_widgets);
-	} else {
-		return -EINVAL;
+		if (!idisp_mask) {
```
@kv2019i Im  abit confused by this check. How will we get here? We already check for codec_count ==2 && idisp_mask on line 128, right?

kv2019i replied:
@ranj063 L128 checks for checks where system has two external codecs: "codec_count == 2 && !idisp_mask". So we reach this block when there are 1 codec and it is an external one.

ranj063 replied:
so this is where we take care of the case where the iDisp codecs are missing but the topology still has the idisp definitions?

ranj063 replied:
and looking at the second commit, the case when codec count is 1 and it is the HDA codec, we should choose the non-idisp topology no? So do we still need the dummy_codec assignments here?

kv2019i replied:
@ranj063 yes -- we basicly have two major DSP topology types for generic HDA. The first if branch is for the idisp-only topology, while the else branch handles the other topology.

ranj063 replied:
ahh so there is no topology that handles just the external HDA codec and no idisp codecs. gotcha!. sorry I got confused there for a bit.

kv2019i replied:
@ranj063 for your second question, yes, it is still needed. we do not have a "external codec only" topology for HDA. so in the case system has only one external codec, we need to use the topology that has idisp nodes, and it goes to the second branch in the machine driver (with dummy codec assigned)

I think we could use this approach in other cases as well to keep the number of distinct topology files reasonable. E.g. define the superset of connections in topology file, and at runtime disable graphs edges that are not connected.

ranj063 replied:
got it! thanks @kv2019i I think the PR looks good now!

kv2019i commented:
@plbossart wrote:
> Should
> a) merge this
> b) send upstream directly
> c) merge upstream stuff then resolve conflicts then send upstream?

Let's take (c). I'll add your reviewed-by tags and send directly. This patch is not critical to have in sof-dev (but is critical for people with affected hw) and I don't have any follow-up patches to these parts of the code, so might as well sent to the list.

kv2019i commented:
Sent to upstream, let's merge this via upstream. Keeping open until merged upstream.

kv2019i commented:
Merged upstream and in sof-dev as well (via rebase).
